### ISSUE TITLE
Race conditions with dependencies

### ISSUE THREAD

**Issue by LaurentGoderre** [wet-boew/wet-boew#7072]
Since WET 4 launched I have notices several instances of race condition that occur with using the `complete` callback of yepnope. It usually harder to reproduce but for application with a lot of scripts processing, the delay incurred by this extra processing can completely break some plugins. The latest instance I discovered is that datatable fails to load because of this line https://github.com/wet-boew/wet-boew/blob/master/src/plugins/tables/tables.js#L89 because while the script has loaded, the plugin is not yet loaded.

It might be necessary to require all lazy-loaded dependency to provide a callback for when the dependency has finished loading, and open PR upstream to add these callback

**Comment by hsrudnicki** [comment]
Please update this issue's status, thx

**Comment by andre-gh** [comment]
Oh wow, you just reminded me, I ran into this back a couple projects ago at my old dept, it was insanely annoying (mostly in IE).  The done callback would fire right away incorrectly. I never figured it out and after pulling my hair out just ended up only initing datatables for the first time after the page load event had occured and that worked around it.

**Comment by andre-gh** [comment]
However, it would only do so some of the time, and with some dependencies, I wrote it up as "possibly an IE bug due to caching when behind HTTP auth"

**Comment by LaurentGoderre** [comment]
The thing is that this not really noticeable in simple content or in example but when within a web app that has several resource to load, the problem occurs more often. The best way to avoid this is by adding a callback to libraries when they are fully initalized

**Comment by andre-gh** [comment]
That seems like working around the problem though, is it an IE bug?  Is it a yupnope bug?

**Comment by LaurentGoderre** [comment]
Nope. Yepnope (and every script loader) tells you when the script is loaded in memory, not when it's done executing. It may appear the same thing but it really isn't. Some of dependency do a lot of processing when it's being run for the first time. The Yepnope callback is not safe in telling you that a library is ready.

For example, sometimes when executing on the Yepnope callback, the jQuery plugin isn't even in `jQuery.fn` yet

**Comment by andre-gh** [comment]
Oh wow then jeepers we have a problem

**Comment by LaurentGoderre** [comment]
It's good design to have a callback so I don't see why any dependency we use would have a problem with using it, and it's so little code.

**Comment by andre-gh** [comment]
Yeah, we should get on evaluating where we depend on this and fixing it.  I knew that script loaders only worked on loading rather than execute, but in my head I saw how many places WET depends on "complete" (including the yupnope docs) being when the script has executed that I started assuming yupnope somehow worked otherwise.

**Comment by pjackson28** [comment]
Some do have the callback, an equivalent event or a DOM update (e.g., adding a class to the HTML element) to indicate it is ready so we could always use those. For instance DataTables has a init.dt event (https://datatables.net/reference/event/). Not as clean as a standardized callback across all library plugins (which will likely never happen) but it is something we could use to address this race condition. Perhaps we could standardize in the WET plugin architecture a way of specifying what callback, event or DOM to wait on before proceeding (so a standardize block that allows the dev to specify what to wait on).

**Comment by andre-gh** [comment]
I am all for standardizing plugins

**Comment by andre-gh** [comment]
http://www.html5rocks.com/en/tutorials/speed/script-loading/ is an amusing read

**Comment by LaurentGoderre** [comment]
@pjackson28 are you sure that init.dt calls when the plugin is initialized and not when an instance is initialized. I find a lot of plugin have events for when instances are ready, but not the library itself.

**Comment by LaurentGoderre** [comment]
@andre-gh good read indeed. I am wondering if we could prototype some stuff, using a script that is slow on purpose to try an find a way.

**Comment by andre-gh** [comment]
I can do all sorts of testing as long as it is IE11 or Firefox :(  I miss having browser VMs.

**Comment by pjackson28** [comment]
@LaurentGoderre Taken from the DataTables event reference for init.dt: "Initialisation complete event - fired when DataTables has been fully initialised and data loaded." which is very different than preInit.dt: "Initialisation started event - triggered immediately before data load.".

**Comment by LaurentGoderre** [comment]
The keyword here is "data loaded". This is not always good enough. What if you need to modify settings before the data loads, or load a plugin for that library before the data is loaded.

What we need is an event for when the library is initiated but before any instance has been created.

**Comment by pjackson28** [comment]
@LaurentGoderre Umm, did you see the part I wrote about preInit.dt ("Initialisation started event - triggered immediately before data load")? If that is not good enough, then there is also processing.dt: "Processing event - fired when DataTables is processing data". It those aren't good enough then I doubt you'll ever get what you are looking for.

**Comment by LaurentGoderre** [comment]
I'm not sure it's early enough because the plugin has already created the instance.

https://datatables.net/reference/event/preInit

We can always propose the callback as a PR.

**Comment by pjackson28** [comment]
Okay, then what about using the existing timerpoke event to fire off the same readiness test over and over until the readiness conditions. For instance, if your condition is a certain object must exist on the window object then on every timerpoke check for the existence before proceeding. That way you can be very specific on the condition and don't have to worry about a separate polling mechanism. Likely much more achievable than getting all the dependencies to implement a very specific callback.

**Comment by LaurentGoderre** [comment]
What about using that as a last resource? Use a callback if it's available, if not, discuss the possibility of a callback with the author and implement the polling in the meantime?

**Comment by pjackson28** [comment]
Considering how long it can take to get changes is upstream libraries (over a year and counting in at least one case) that would likely be a prudent approach. 

So is it agreed that the way forward is to implement a standard code block in plugins dedicated to confirming the readiness of dependencies before proceeding (to make up for Yepnope complete being insufficient) by doing the following?
1. Waiting on an existing callback or event from the dependencies.
2. Failing that, implementing the timepoke polling readiness checks and at the same time try to get an appropriate callback or event in place upstream for the dependencies.

**Comment by LaurentGoderre** [comment]
I think that is our best bet yes

**Comment by hsrudnicki** [comment]
Based on the above, can be this issue, just labeled "Going Stale", closed as resolved or otherwise?
Either way, please update this issue's status, thank you
